File: 005.LongestPalindromicSubstring/Palindrome.java
class Solution {
    public String longestPalindrome(String s) {
        int maxLen = 1;
        String longestPs = s.substring(0, 1);

        for (int i = 0; i < s.length(); i++) {
            String psOdd = extendPalindrome(s, i, i);
            String psEven = extendPalindrome(s, i, i+1);
            String ps = psOdd.length() > psEven.length() ? psOdd : psEven;
            if (ps.length() > maxLen) {
                maxLen = ps.length();
                longestPs = ps;
            }
        }

        return longestPs;
    }

    public String extendPalindrome(String s, int i, int j) {
        while (i >= 0 && j < s.length() && s.charAt(i) == s.charAt(j)) {
            i--; j++;
        }
        return s.substring(i+1, j);
    }
}

File: 005.LongestPalindromicSubstring/palindrome.cc
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        int low = 0, maxLen = 1;

        for (int i = 0; i < int(s.length()); i++) {
            extendPalindrome(s, i, i, low, maxLen);
            extendPalindrome(s, i, i+1, low, maxLen);
        }

        return s.substr(low, maxLen);
    }

    void extendPalindrome(string s, int i, int j, int &low, int &maxLen) {
        while (i >= 0 && j < int(s.length()) && s[i] == s[j]) {
            i--; j++;
        }
        if (j - i - 1 > maxLen) {
            maxLen = j - i - 1;
            low = i + 1;
        }
    }
};

int main(int argc, char const *argv[]) {
    Solution sl = Solution();
    string s = "xyzabcbaqqq";
    cout << sl.longestPalindrome(s) << endl;
    return 0;
}
